Cannot build with clang-11 and libc++ in C++20 mode.

This component fails to compile due to not using `std::allocator_traits` with clang-11 and libc++ in C++20 mode.

```
boost/wave/util/flex_string.hpp:547:25: error: no type named 'size_type' in 'std::__1::allocator<char>'
    typedef typename A::size_type size_type;
            ~~~~~~~~~~~~^~~~~~~~~
```

* [build log](https://github.com/boostorg/wave/files/4447651/build-log.txt)
* [config.jam](https://github.com/boostorg/wave/files/4447652/config.jam.txt)

Command used to build boost:

```
b2 --layout=system
   --project-config=/mnt/c/Workspace/vcpkg/buildtrees/boost/x64-linux-dbg/config.jam
   --prefix=/mnt/c/Workspace/vcpkg/packages/boost_x64-linux/debug
   --libdir=/mnt/c/Workspace/vcpkg/packages/boost_x64-linux/debug/lib
   --exec-prefix=/mnt/c/Workspace/vcpkg/packages/boost_x64-linux/debug/bin
   --includedir=/mnt/c/Workspace/vcpkg/packages/boost_x64-linux/debug/../include
   --cmakedir=/mnt/c/Workspace/vcpkg/packages/boost_x64-linux/debug/share/boost
   --build-dir=/mnt/c/Workspace/vcpkg/buildtrees/boost/x64-linux-dbg/build
   --without-graph_parallel --without-mpi --without-python
   --includedir=/mnt/c/Workspace/vcpkg/packages/boost_x64-linux/include
   architecture=x86 address-model=64
   variant=debug threading=multi
   link=static runtime-link=shared
   debug-symbols=on install
```

The `log` component has a similar issue: https://github.com/boostorg/log/issues/114

--- a/include/boost/wave/util/flex_string.hpp
+++ b/include/boost/wave/util/flex_string.hpp
@@ -287,29 +287,31 @@ inline bool operator!=(const mallocator<T>&,
   return false;
 }
 
-template <class Allocator>
-typename Allocator::pointer Reallocate(
-    Allocator& alloc,
-    typename Allocator::pointer p,
-    typename Allocator::size_type oldObjCount,
-    typename Allocator::size_type newObjCount,
-    void*)
-{
-    // @@@ not implemented
-    return NULL;
-}
+#if defined(BOOST_NO_CXX11_ALLOCATOR)
+template<class A>
+struct allocator_traits {
+    typedef typename A::value_type value_type;
+    typedef typename A::pointer pointer;
+    typedef typename A::const_pointer const_pointer;
+    typedef typename A::size_type size_type;
+    typedef typename A::reference reference;
+    typedef typename A::const_reference const_reference;
 
-template <class Allocator>
-typename Allocator::pointer Reallocate(
-    Allocator& alloc,
-    typename Allocator::pointer p,
-    typename Allocator::size_type oldObjCount,
-    typename Allocator::size_type newObjCount,
-    mallocator<void>*)
-{
-    // @@@ not implemented
-    return NULL;
-}
+    static pointer allocate( A& a, size_type n, const_pointer hint )
+    {
+        return a.allocate(n, hint);
+    }
+
+    static void deallocate( A& a, pointer p, size_type n )
+    {
+        a.deallocate(p, n);
+    }
+
+    static size_type max_size( A const& a ) { return a.max_size(); }
+};
+#else
+using std::allocator_traits;
+#endif
 
 ////////////////////////////////////////////////////////////////////////////////
 // class template SimpleStringStorage
@@ -331,11 +333,7 @@ public:
     };
     static const Data emptyString_;
 
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-    typedef typename A::size_type size_type;
-#else
-    typedef typename std::allocator_traits<A>::size_type size_type;
-#endif
+    typedef typename allocator_traits<A>::size_type size_type;
 
 private:
     Data* pData_;
@@ -548,22 +546,13 @@ SimpleStringStorage<E, A>::emptyString_ =
 template <typename E, class A = std::allocator<E> >
 class AllocatorStringStorage : public A
 {
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-    typedef typename A::size_type size_type;
-#else
-    typedef typename std::allocator_traits<A>::size_type size_type;
-#endif
+    typedef typename allocator_traits<A>::size_type size_type;
     typedef typename SimpleStringStorage<E, A>::Data Data;
 
     void* Alloc(size_type sz, const void* p = 0)
     {
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-        return A::allocate(1 + (sz - 1) / sizeof(E),
+        return allocator_traits<A>::allocate(*this, 1 + (sz - 1) / sizeof(E),
             static_cast<const char*>(p));
-#else
-        return std::allocator_traits<A>::allocate(*this, 1 + (sz - 1) / sizeof(E),
-            static_cast<const char*>(p));
-#endif
     }
 
     void* Realloc(void* p, size_type oldSz, size_type newSz)
@@ -576,11 +565,7 @@ class AllocatorStringStorage : public A
 
     void Free(void* p, size_type sz)
     {
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-        A::deallocate(static_cast<E*>(p), sz);
-#else
-        std::allocator_traits<A>::deallocate(*this, static_cast<E*>(p), sz);
-#endif
+        allocator_traits<A>::deallocate(*this, static_cast<E*>(p), sz);
     }
 
     Data* pData_;
@@ -683,11 +668,7 @@ public:
     { return size_type(end() - begin()); }
 
     size_type max_size() const
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-    { return A::max_size(); }
-#else
-    { return std::allocator_traits<A>::max_size(*this); }
-#endif
+    { return allocator_traits<A>::max_size(*this); }
 
     size_type capacity() const
     { return size_type(pData_->pEndOfMem_ - pData_->buffer_); }
@@ -778,11 +759,7 @@ public: // protected:
     typedef typename base::iterator iterator;
     typedef typename base::const_iterator const_iterator;
     typedef A allocator_type;
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-    typedef typename A::size_type size_type;
-#else
-    typedef typename std::allocator_traits<A>::size_type size_type;
-#endif
+    typedef typename allocator_traits<A>::size_type size_type;
 
     VectorStringStorage(const VectorStringStorage& s) : base(s)
     { }
@@ -899,11 +876,7 @@ public:
     typedef value_type* iterator;
     typedef const value_type* const_iterator;
     typedef typename Storage::allocator_type allocator_type;
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-    typedef typename allocator_type::size_type size_type;
-#else
-    typedef typename std::allocator_traits<allocator_type>::size_type size_type;
-#endif
+    typedef typename allocator_traits<allocator_type>::size_type size_type;
 
 private:
   enum { temp1 = threshold * sizeof(value_type) > sizeof(Storage)
@@ -1230,11 +1203,7 @@ public:
     typedef typename Storage::iterator iterator;
     typedef typename Storage::const_iterator const_iterator;
     typedef typename Storage::allocator_type allocator_type;
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-    typedef typename allocator_type::size_type size_type;
-#else
-    typedef typename std::allocator_traits<allocator_type>::size_type size_type;
-#endif
+    typedef typename allocator_traits<allocator_type>::size_type size_type;
     typedef typename Storage::value_type& reference;
 
 private:
@@ -1495,19 +1464,11 @@ public:
     typedef typename traits_type::char_type value_type;
     typedef A allocator_type;
 
-#if defined(BOOST_NO_CXX11_ALLOCATOR)
-    typedef typename A::reference reference;
-    typedef typename A::const_reference const_reference;
-    typedef typename A::pointer pointer;
-    typedef typename A::const_pointer const_pointer;
-    typedef typename A::size_type size_type;
-#else
-    typedef typename std::allocator_traits<A>::value_type& reference;
-    typedef typename std::allocator_traits<A>::value_type const& const_reference;
-    typedef typename std::allocator_traits<A>::pointer pointer;
-    typedef typename std::allocator_traits<A>::const_pointer const_pointer;
-    typedef typename std::allocator_traits<A>::size_type size_type;
-#endif
+    typedef typename allocator_traits<A>::value_type& reference;
+    typedef typename allocator_traits<A>::value_type const& const_reference;
+    typedef typename allocator_traits<A>::pointer pointer;
+    typedef typename allocator_traits<A>::const_pointer const_pointer;
+    typedef typename allocator_traits<A>::size_type size_type;
 
     typedef typename Storage::iterator iterator;
     typedef typename Storage::const_iterator const_iterator;

--- a/include/boost/wave/util/flex_string.hpp
+++ b/include/boost/wave/util/flex_string.hpp
@@ -331,7 +331,11 @@ public:
     };
     static const Data emptyString_;
 
+#if defined(BOOST_NO_CXX11_ALLOCATOR)
     typedef typename A::size_type size_type;
+#else
+    typedef typename std::allocator_traits<A>::size_type size_type;
+#endif
 
 private:
     Data* pData_;
@@ -544,7 +548,11 @@ SimpleStringStorage<E, A>::emptyString_ =
 template <typename E, class A = std::allocator<E> >
 class AllocatorStringStorage : public A
 {
+#if defined(BOOST_NO_CXX11_ALLOCATOR)
     typedef typename A::size_type size_type;
+#else
+    typedef typename std::allocator_traits<A>::size_type size_type;
+#endif
     typedef typename SimpleStringStorage<E, A>::Data Data;
 
     void* Alloc(size_type sz, const void* p = 0)
@@ -568,7 +576,11 @@ class AllocatorStringStorage : public A
 
     void Free(void* p, size_type sz)
     {
+#if defined(BOOST_NO_CXX11_ALLOCATOR)
         A::deallocate(static_cast<E*>(p), sz);
+#else
+        std::allocator_traits<A>::deallocate(*this, static_cast<E*>(p), sz);
+#endif
     }
 
     Data* pData_;
@@ -766,7 +778,11 @@ public: // protected:
     typedef typename base::iterator iterator;
     typedef typename base::const_iterator const_iterator;
     typedef A allocator_type;
+#if defined(BOOST_NO_CXX11_ALLOCATOR)
     typedef typename A::size_type size_type;
+#else
+    typedef typename std::allocator_traits<A>::size_type size_type;
+#endif
 
     VectorStringStorage(const VectorStringStorage& s) : base(s)
     { }
@@ -1470,19 +1486,19 @@ public:
     typedef T traits_type;
     typedef typename traits_type::char_type value_type;
     typedef A allocator_type;
-    typedef typename A::size_type size_type;
-    typedef typename A::difference_type difference_type;
 
 #if defined(BOOST_NO_CXX11_ALLOCATOR)
     typedef typename A::reference reference;
     typedef typename A::const_reference const_reference;
     typedef typename A::pointer pointer;
     typedef typename A::const_pointer const_pointer;
+    typedef typename A::size_type size_type;
 #else
     typedef typename std::allocator_traits<A>::value_type& reference;
     typedef typename std::allocator_traits<A>::value_type const& const_reference;
     typedef typename std::allocator_traits<A>::pointer pointer;
     typedef typename std::allocator_traits<A>::const_pointer const_pointer;
+    typedef typename std::allocator_traits<A>::size_type size_type;
 #endif
 
     typedef typename Storage::iterator iterator;

--- a/include/boost/wave/util/flex_string.hpp
+++ b/include/boost/wave/util/flex_string.hpp
@@ -899,7 +899,11 @@ public:
     typedef value_type* iterator;
     typedef const value_type* const_iterator;
     typedef typename Storage::allocator_type allocator_type;
+#if defined(BOOST_NO_CXX11_ALLOCATOR)
     typedef typename allocator_type::size_type size_type;
+#else
+    typedef typename std::allocator_traits<allocator_type>::size_type size_type;
+#endif
 
 private:
   enum { temp1 = threshold * sizeof(value_type) > sizeof(Storage)
@@ -1226,7 +1230,11 @@ public:
     typedef typename Storage::iterator iterator;
     typedef typename Storage::const_iterator const_iterator;
     typedef typename Storage::allocator_type allocator_type;
+#if defined(BOOST_NO_CXX11_ALLOCATOR)
     typedef typename allocator_type::size_type size_type;
+#else
+    typedef typename std::allocator_traits<allocator_type>::size_type size_type;
+#endif
     typedef typename Storage::value_type& reference;
 
 private:

